elyra/util/git.py

```diff
@@ -0,0 +1,84 @@
+#
+# Copyright 2018-2021 Elyra Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from github import Github, GithubException
+from traitlets.config import LoggingConfigurable
+from typing import Optional
+
+
+class GithubClient(LoggingConfigurable):
+
+    def __init__(self,
+                 token: str,
+                 repo: str,
+                 branch: Optional[str] = None,
+                 server_url: Optional[str] = None):
```

Comment:
Suggestion for a slight tweak. By setting the default here

```suggestion
                 server_url: Optional[str] = 'https://api.github.com'):
```

the code below
```
        if server_url:
            self.client = Github(login_or_token=token, base_url=server_url)
        else:
            self.client = Github(login_or_token=token)
```

could be simplified to

```
        self.client = Github(login_or_token=token, base_url=server_url)
```

Reply:
I would prefer we only define the default in one place - either in the schema or as an env-accessible value that is only used here.  Having it defined in multiple locations is confusing to maintainers.

If we go with the schema default (which feels more correct) all instances will have at least that value - i.e., it will be tantamount to a required property.  If we take that route, and all instances of `GithubClient` derive the parameters of metadata instances, then `server_url` would be a required value as well.

So I guess it would boil down to is _where is the default value defined?_ and I would say it's inconsistent to not define this in the schema (although less flexible as well).  (Note that we could address this via the metadata approach by implementing a class for Runtime instances where _that_ implementation deals with configuration, env, etc.)

Reply:
> either in the schema or as an env-accessible value that is only used here

I just re-read why Patrick suggested the default in the schema (being accessible from the UI), so would retract the latter-portion of my previous statement and preferring we go with the schema default.